scripts/gen_visitors.py

```diff
@@ -0,0 +1,293 @@
+    child_method: str,
+) -> str:
+    # Helper to emit a for-loop over a collection with optional headers guard
+    if field_name == "headers":
+        return f"""\
+        if not self.skip_headers:
+            for v in {iter_expr}:
+                await self._visit_{child_method}(fs, v)"""
+    else:
+        return f"""\
+        for v in {iter_expr}:
+            await self._visit_{child_method}(fs, v)"""
+
+
+def emit_singular(
+    field_name: str, access_expr: str, child_method: str, check_presence: bool
+) -> str:
+    # Helper to emit a singular field visit with presence check and optional headers guard
+    if check_presence:
+        if field_name == "headers":
+            return f"""\
+        if not self.skip_headers:
+            if o.HasField("{field_name}"):
+                await self._visit_{child_method}(fs, {access_expr})"""
+        else:
+            return f"""\
+        if o.HasField("{field_name}"):
+            await self._visit_{child_method}(fs, {access_expr})"""
+    else:
+        if field_name == "headers":
+            return f"""\
+        if not self.skip_headers:
+            await self._visit_{child_method}(fs, {access_expr})"""
+        else:
+            return f"""\
+        await self._visit_{child_method}(fs, {access_expr})"""
+
+
+class VisitorGenerator:
+    def __init__(self):
+        # Track which message descriptors have visitor methods generated
+        self.generated: dict[str, bool] = {
+            Payload.DESCRIPTOR.full_name: True,
+            Payloads.DESCRIPTOR.full_name: True,
+        }
+        self.in_progress: set[str] = set()
+        self.methods: list[str] = [
+            """    async def _visit_temporal_api_common_v1_Payload(self, fs, o):
+            await fs.visit_payload(o)
+    """,
+            """    async def _visit_temporal_api_common_v1_Payloads(self, fs, o):
+            await fs.visit_payloads(o.payloads)
+    """,
+            """    async def _visit_payload_container(self, fs, o):
+            await fs.visit_payloads(o)
+    """,
+        ]
+
+    def check_repeated(self, child_desc, field, iter_expr) -> Optional[str]:
+        # Special case for repeated payloads, handle them directly
+        if child_desc.full_name == Payload.DESCRIPTOR.full_name:
+            return emit_singular(field.name, iter_expr, "payload_container", False)
+        else:
+            child_needed = self.walk(child_desc)
+            if child_needed:
+                return emit_loop(
+                    field.name,
+                    iter_expr,
+                    name_for(child_desc),
+                )
+            else:
+                return None
+
+    def walk(self, desc: Descriptor) -> bool:
+        key = desc.full_name
+        if key in self.generated:
+            return self.generated[key]
+        if key in self.in_progress:
+            # Break cycles; if another path proves this node needed, we'll revisit
+            return False
+
+        needed = False
```

Comment:
Could you add a comment explaining this `needed` variable since it plays an important role but isn't obvious what it means?

Reply:
I wonder if `has_payload` would be clearer.